### PR TITLE
Feature 2289 review assignments

### PR DESCRIPTION
**Pull request by mvolkmann**

### CODE REVIEW

File: web-ui/src/components/member_selector/member_selector_dialog/MemberSelectorDialog.jsx [hunk 1]
```diff
@@ -165,7 +165,7 @@ const MemberSelectorDialog = ({
         );
       }
     }
-  }, [open]);
+  }, [open, selectedMembers]);
```
Comment (mvolkmann) [inline web-ui/src/components/member_selector/member_selector_dialog/MemberSelectorDialog.jsx:168]: The `initializeChecked` function called in this `useEffect` uses the value of  `selectedMembers`.

File: web-ui/src/components/reviews/TeamReviews.css [hunk 2]
```diff
@@ -0,0 +1,13 @@
+.team-reviews {
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.css:1]: This styles the rows of members and their reviewers.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 3]
```diff
@@ -87,14 +87,10 @@ import TeamMemberReview from './TeamMemberReview';
 
 import DatePickerField from './periods/DatePickerField.jsx';
 import './periods/DatePickerField.css';
+import './TeamReviews.css';
 
 const propTypes = {
-  teamMembers: PropTypes.arrayOf(
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:92]: This was not one of the props being used.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 4]
```diff
@@ -87,14 +87,10 @@ import TeamMemberReview from './TeamMemberReview';
...
-    PropTypes.shape({
-      id: PropTypes.string,
-      onBack: PropTypes.func
-    })
-  ),
+  onBack: PropTypes.func,
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:93]: This is called when users click the "Request Approval" button.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 5]
```diff
@@ -141,14 +137,17 @@ const TeamReviews = ({ onBack, periodId }) => {
   const history = useHistory();
   const location = useLocation();
 
+  const [assignments, setAssignments] = useState([]);
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:140]: This holds all the assignments (member <-> reviewer) for the current review period.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 6]
```diff
@@ -141,14 +137,17 @@ const TeamReviews = ({ onBack, periodId }) => {
...
   const [confirmOpen, setConfirmOpen] = useState(false);
   const [includeAll, setIncludeAll] = useState(false);
   const [memberFilters, setMemberFilters] = useState([]);
   const [memberSelectorOpen, setMemberSelectorOpen] = useState(false);
   const [newRequestOpen, setNewRequestOpen] = useState(false);
   const [query, setQuery] = useState({});
+  const [reviewerSelectorOpen, setReviewerSelectorOpen] = useState(false);
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:147]: This controls whether a `MemberSelectorDialog` is open.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 7]
```diff
@@ -12,46 +12,26 @@ import React, {
 } from 'react';
 import { useLocation, useHistory } from 'react-router-dom';
 
+import { AddCircle, Archive, Delete, Unarchive } from '@mui/icons-material';
 import {
-  AddCircle,
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:16]: I removed a lot of code that is no longer used.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 8]
```diff
@@ -141,46 +112,49 @@ const TeamReviews = ({ onBack, periodId }) => {
   const history = useHistory();
   const location = useLocation();
 
+  const [assignments, setAssignments] = useState([]);
   const [confirmOpen, setConfirmOpen] = useState(false);
   const [includeAll, setIncludeAll] = useState(false);
   const [memberFilters, setMemberFilters] = useState([]);
   const [memberSelectorOpen, setMemberSelectorOpen] = useState(false);
-  const [newRequestOpen, setNewRequestOpen] = useState(false);
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:148]: No longer used.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 9]
```diff
@@ -141,46 +112,49 @@ const TeamReviews = ({ onBack, periodId }) => {
...
   const [query, setQuery] = useState({});
+  const [reviewerSelectorOpen, setReviewerSelectorOpen] = useState(false);
   const [reviews, setReviews] = useState(null);
   const [selectedMember, setSelectedMember] = useState(null);
+  const [selectedReviewers, setSelectedReviewers] = useState([]);
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:124]: This is used by the `MemberSelectorDialog`.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 10]
```diff
@@ -141,46 +112,49 @@ const TeamReviews = ({ onBack, periodId }) => {
...
   const [selfReviews, setSelfReviews] = useState({});
   const [teamMembers, setTeamMembers] = useState([]);
   const [toDelete, setToDelete] = useState(null);
 
-  const creatingReview = useRef(false);
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:156]: No longer used.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 11]
```diff
@@ -141,46 +112,49 @@ const TeamReviews = ({ onBack, periodId }) => {
...
   const loadedReviews = useRef(false);
   const loadingReviews = useRef(false);
 
   const csrf = selectCsrfToken(state);
   const currentMembers = selectCurrentMembers(state);
+  const memberMap = currentMembers.reduce((map, member) => {
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:134]: This allows quickly finding a member based on their id.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 12]
```diff
@@ -141,46 +112,49 @@ const TeamReviews = ({ onBack, periodId }) => {
...
+    map[member.id] = member;
+    return map;
+  }, {});
   const currentUser = selectCurrentUser(state);
   const isAdmin = selectIsAdmin(state);
-  const myTeam = selectMyTeam(state);
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:164]: Lots more unused code was removed.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 13]
```diff
@@ -141,46 +112,49 @@ const TeamReviews = ({ onBack, periodId }) => {
...
-    [setNewRequestOpen]
-  );
 
   const reviewAssignmentsUrl = '/services/review-assignments';
 
   useEffect(() => {
     loadTeamMembers();
   }, [currentMembers]);
 
+  useEffect(() => {
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:148]: When the circled plus icon for a member is clicked, this gets the reviewers that are currently assigned to that member so they can be passed to the `MemberSelectorDialog`.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 14]
```diff
@@ -141,46 +112,49 @@ const TeamReviews = ({ onBack, periodId }) => {
...
+    const reviewers = selectedMember ? getReviewers(selectedMember) : [];
+    setSelectedReviewers(reviewers);
+  }, [selectedMember]);
+
+  const editReviewers = member => {
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:153]: This is called when the circled plus icon for a member is clicked.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 15]
```diff
@@ -393,6 +308,21 @@ const TeamReviews = ({ onBack, periodId }) => {
     history.goBack();
   }, [csrf, dispatch, toDelete, handleConfirmClose]);
 
+  const getReviewers = useCallback(
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:311]: This is called by the `editReviewers` function.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 16]
```diff
@@ -592,105 +504,131 @@ const TeamReviews = ({ onBack, periodId }) => {
     loadReviews();
   }, [loadReviews]);
 
-  const reloadReviews = useCallback(() => {
-    loadedReviews.current = false;
-    loadReviews();
-  }, [loadReviews]);
+  const deleteReviewer = async (member, reviewer) => {
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:507]: This is called when the "x" in a reviewer `Chip` is clicked to review the corresponding reviewer from the member described by that row.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 17]
```diff
@@ -592,105 +504,131 @@ const TeamReviews = ({ onBack, periodId }) => {
...
-          }
+      });
+      if (res.error) throw new Error(res.error.message);
+      onBack();
+    } catch (err) {
+      console.error('TeamReviews.jsx deleteReviewer:', err);
+    }
+  };
+
+  const compareStrings = (s1, s2) => (s1 || '').localeCompare(s2 || '');
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:552]: This is used by the `sortMembers` function below.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 18]
```diff
@@ -592,105 +504,131 @@ const TeamReviews = ({ onBack, periodId }) => {
...
+
+  const sortMembers = members =>
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:554]: This is used to sort the list of reviewers for each member.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 19]
```diff
@@ -592,105 +504,131 @@ const TeamReviews = ({ onBack, periodId }) => {
...
+    members.sort((a, b) => {
+      let compare = compareStrings(a.lastName, b.lastName);
+      if (compare === 0) compare = compareStrings(a.firstName, b.firstName);
+      return compare;
+    });
+
+  const updateReviewers = async (member, reviewers) => {
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:561]: This is called when the SAVE button in the `MemberSelectorDialog` is clicked.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 20]
```diff
@@ -592,105 +504,131 @@ const TeamReviews = ({ onBack, periodId }) => {
...
+  };
 
-  return (
-    <Root>
-      <Tooltip title="Back">
-        <Button onClick={onBack} aria-label="Back">
-          <ArrowBack /> Back
-        </Button>
-      </Tooltip>
+  const closeReviewerDialog = () => {
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:603]: This is called when the "X" button in the upper-left of the `MemberSelectorDialog` is clicked.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 21]
```diff
@@ -592,105 +504,131 @@ const TeamReviews = ({ onBack, periodId }) => {
...
+    setSelectedMember(null);
+    setSelectedReviewers([]);
+    setReviewerSelectorOpen(false);
+  };
 
+  const REVIEWER_LIMIT = 2;
+  const renderReviewers = member => {
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:610]: This renders chips for the reviewers of a  given member. Currently a maximum of two are displayed. If their are more, text following the chips indicates the number of reviewers that are not shown.

File: web-ui/src/components/reviews/TeamReviews.jsx [hunk 22]
```diff
@@ -778,45 +702,54 @@ const TeamReviews = ({ onBack, periodId }) => {
...
-                animation="wave"
-                variant="circular"
-                width={40}
-                height={40}
-              />
-            </ListItemAvatar>
+
+      <List dense role="list" sx={{ height: '50%', overflowY: 'scroll' }}>
+        {teamMembers.map(member => (
+          <ListItem
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/TeamReviews.jsx:708]: This renders a row for one member.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 23]
```diff
@@ -14,7 +14,6 @@ import {
   Box,
   Button,
   FormControl,
-  IconButton,
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:17]: No longer used.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 24]
```diff
@@ -25,7 +24,6 @@ import {
   Select,
   Skeleton,
   TextField,
-  Tooltip,
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:28]: No longer used.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 25]
```diff
@@ -60,7 +58,6 @@ const propTypes = {
   onSelect: PropTypes.func
 };
 const displayName = 'ReviewPeriods';
-const feedbackTemplateUrl = '/services/feedback/templates';
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:63]: No longer used.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 26]
```diff
@@ -347,8 +344,27 @@ const ReviewPeriods = ({ onPeriodSelected, mode }) => {
 
   const onPeriodClick = useCallback(
     id => {
-      if (selectReviewPeriod(state, id)?.reviewStatus === ReviewStatus.OPEN) {
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:350]: We are now handling more status values than OPEN.

File: web-ui/src/components/reviews/periods/ReviewPeriods.jsx [hunk 27]
```diff
@@ -409,41 +425,29 @@ const ReviewPeriods = ({ onPeriodSelected, mode }) => {
...
                   ? -1
                   : 1;
             })
-            .map(
-              (
-                {
-                  name,
-                  reviewStatus,
-                  id,
-                  launchDate,
```
Comment (mvolkmann) [inline web-ui/src/components/reviews/periods/ReviewPeriods.jsx:423]: The next three properties are no longer used. Removing them causes several lines to be reformatted by Prettier.

File: web-ui/src/helpers/strings.js [hunk 28]
```diff
@@ -1,5 +1,7 @@
 export const titleCase = text =>
-  text.replace(
-    /\w\S*/g,
-    word => word.charAt(0).toUpperCase() + word.substr(1).toLowerCase()
-  );
+  text
+    .replace(/_/g, ' ')
```
Comment (mvolkmann) [inline web-ui/src/helpers/strings.js:3]: I modified this to replace underscores with spaces.

File: web-ui/src/pages/ReviewsPage.jsx [hunk 29]
```diff
@@ -92,7 +92,7 @@ const ReviewPage = () => {
           <ReviewPeriods onPeriodSelected={onPeriodSelected} />
         ) : (
           <TeamReviews
-            onBack={() => onPeriodSelected(null)}
+            onBack={() => setSelectedPeriod(null)}
```
Comment (mvolkmann) [inline web-ui/src/pages/ReviewsPage.jsx:95]: I corrected the function name.

**Review by jackkeller — APPROVED** [review]
Approved!

**Review by vhscom — APPROVED** [review]
Assuming the `TODO` is for a later pull. Looks good!

**Pull request merged**